### PR TITLE
[sdk-metrics] Turn exemplars on by default in prerelease builds

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Set the default `ExemplarFilterType` to `TraceBased` in prerelease builds to match [spec](https://github.com/open-telemetry/opentelemetry-specification/pull/3994).

### Counters (SimpleFixedSizeExemplarReservoir)

Using `TraceBased` WITHOUT an active trace has some cost (we need to check `Activity.Current.Recorded` in the hot path):

| Method                    | AggregationTemporality | ExemplarFilterType | Mean     | Error    | StdDev   |
|-------------------------- |----------------------- |------------------- |---------:|---------:|---------:|
| CounterHotPath            | Cumulative             | AlwaysOff          | 10.51 ns | 0.150 ns | 0.140 ns |
| CounterWith1LabelsHotPath | Cumulative             | AlwaysOff          | 36.37 ns | 0.253 ns | 0.236 ns |
| CounterWith2LabelsHotPath | Cumulative             | AlwaysOff          | 44.98 ns | 0.709 ns | 0.663 ns |
| CounterWith3LabelsHotPath | Cumulative             | AlwaysOff          | 61.44 ns | 0.963 ns | 0.853 ns |
| CounterHotPath            | Cumulative             | TraceBased         | 11.29 ns | 0.122 ns | 0.114 ns |
| CounterWith1LabelsHotPath | Cumulative             | TraceBased         | 37.48 ns | 0.768 ns | 0.681 ns |
| CounterWith2LabelsHotPath | Cumulative             | TraceBased         | 46.68 ns | 0.943 ns | 0.882 ns |
| CounterWith3LabelsHotPath | Cumulative             | TraceBased         | 64.76 ns | 0.661 ns | 0.586 ns |

Using `TraceBased` WITH an active trace has more cost (we need to check `Activity.Current.Recorded` and do a random-based sample in the hot path):

| Method                    | AggregationTemporality | ExemplarFilterType | Mean     | Error    | StdDev   |
|-------------------------- |----------------------- |------------------- |---------:|---------:|---------:|
| CounterHotPath            | Cumulative             | AlwaysOff          | 10.44 ns | 0.085 ns | 0.080 ns |
| CounterWith1LabelsHotPath | Cumulative             | AlwaysOff          | 36.86 ns | 0.495 ns | 0.439 ns |
| CounterWith2LabelsHotPath | Cumulative             | AlwaysOff          | 45.82 ns | 0.709 ns | 0.628 ns |
| CounterWith3LabelsHotPath | Cumulative             | AlwaysOff          | 61.02 ns | 0.754 ns | 0.705 ns |
| CounterHotPath            | Cumulative             | TraceBased         | 18.32 ns | 0.220 ns | 0.184 ns |
| CounterWith1LabelsHotPath | Cumulative             | TraceBased         | 46.46 ns | 0.778 ns | 0.727 ns |
| CounterWith2LabelsHotPath | Cumulative             | TraceBased         | 53.52 ns | 0.398 ns | 0.353 ns |
| CounterWith3LabelsHotPath | Cumulative             | TraceBased         | 70.00 ns | 0.892 ns | 0.834 ns |

### Histograms (AlignedHistogramBucketExemplarReservoir)

Using `TraceBased` WITHOUT an active trace is interesting. Sometimes I run it things show faster, sometimes it shows slower, and sometimes it shows mixed. I take this as statistically no difference. The cost of the check for `Activity.Current.Recorded` is dwarfed by the other work to find the bucket and do all the updating:

| Method                      | BoundCount | ExemplarFilterType | Mean      | Error    | StdDev   |
|---------------------------- |----------- |------------------- |----------:|---------:|---------:|
| HistogramHotPath            | 10         | AlwaysOff          |  37.37 ns | 0.404 ns | 0.378 ns |
| HistogramWith1LabelHotPath  | 10         | AlwaysOff          |  65.99 ns | 0.685 ns | 0.572 ns |
| HistogramWith3LabelsHotPath | 10         | AlwaysOff          | 110.73 ns | 1.210 ns | 1.132 ns |
| HistogramHotPath            | 10         | TraceBased         |  36.99 ns | 0.348 ns | 0.291 ns |
| HistogramWith1LabelHotPath  | 10         | TraceBased         |  66.86 ns | 1.315 ns | 1.407 ns |
| HistogramWith3LabelsHotPath | 10         | TraceBased         | 113.68 ns | 1.008 ns | 0.894 ns |

Using `TraceBased` WITH an active trace has a lot of cost (we need to check `Activity.Current.Recorded` and we always update exemplar for every measurement in the hot path):

| Method                      | BoundCount | ExemplarFilterType | Mean      | Error    | StdDev   |
|---------------------------- |----------- |------------------- |----------:|---------:|---------:|
| HistogramHotPath            | 10         | AlwaysOff          |  39.93 ns | 0.468 ns | 0.438 ns |
| HistogramWith1LabelHotPath  | 10         | AlwaysOff          |  71.05 ns | 0.947 ns | 0.886 ns |
| HistogramWith3LabelsHotPath | 10         | AlwaysOff          | 109.03 ns | 2.202 ns | 3.799 ns |
| HistogramHotPath            | 10         | TraceBased         |  68.96 ns | 0.863 ns | 0.765 ns |
| HistogramWith1LabelHotPath  | 10         | TraceBased         |  96.48 ns | 0.956 ns | 0.894 ns |
| HistogramWith3LabelsHotPath | 10         | TraceBased         | 153.84 ns | 1.680 ns | 1.489 ns |

This is an interesting area @cijothomas and I have discussed. The spec says for `AlignedHistogramBucketExemplarReservoir` always keep the last exemplar seen for a bucket. There's a lot of overriding as a result (wasted cycles). A simple thing to do would be keep only the first exemplar for a given export. Or do something more like `SimpleFixedSizeExemplarReservoir` where we always keep the first one then randomly decide whether or not to keep subsequent exemplars 🤔

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,11 @@
   function when configuring a view (applies to individual metrics).
   ([#5542](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5542))
 
+* **Experimental (pre-release builds only):** The default `ExemplarFilterType`
```
Comment (vishweshbankwar) [inline src/OpenTelemetry/CHANGELOG.md:15]: Could we reword this to say something like "Exemplars will be enabled automatically with default filter type set to TraceBased and can be changed by calling SetExemplarFilter on MeterProvider".

**Review by vishweshbankwar — APPROVED** [review]
LGTM - left a suggestion for changelog.

**Review by cijothomas — APPROVED** [review]
I am okay with this change, but I want to see if we can make a NoopExemplarReservoir, and make it as the default for non-histograms (spec is flexible to allow that) in the 1st stable release.

File: src/OpenTelemetry/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,11 @@
...
+  on `MeterProvider` is now `ExemplarFilterType.TraceBased` which will enable
```
Comment (reyang) [inline src/OpenTelemetry/CHANGELOG.md:16]: What's the perf implications that the users should be aware of?

File: src/OpenTelemetry/CHANGELOG.md (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/CHANGELOG.md]: I tweaked it a bit

File: src/OpenTelemetry/CHANGELOG.md (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/CHANGELOG.md]: Let's wait on this. My guess is most users will skip over anything prefixed with `**Experimental (pre-release builds only):**` in the CHANGELOG. What I think would be more useful is on the final entry where we make everything public for stable builds we can add a link there to something in the docs. Thinking like: `Understanding performance implications when sampling Exemplars`.

Comment (reyang) [reply src/OpenTelemetry/CHANGELOG.md]: I'm fine to address the changelog/doc later.

I think we still need to know the perf implication of this PR as it serves as a critical input while making decisions.

Comment (CodeBlanch) [reply src/OpenTelemetry/CHANGELOG.md]: Updated the description with some benchmarks.

Comment (utpilla) [reply src/OpenTelemetry/CHANGELOG.md]: Could you also provide the increase in memory consumption per metric since we would now allocate an `ExemplarReservoir` instance for each MetricPoint?

**Comment by cijothomas** [comment]
> I am okay with this change, but I want to see if we can make a NoopExemplarReservoir, and make it as the default for non-histograms (spec is flexible to allow that) in the 1st stable release.

After looking at the perf numbers, the overhead is non-trivial. So I recommend to keep it off by default for every metric. Users can opt-in to each metric (using views). (Not many backends/venodors are known to support exemplars.)

**Comment by CodeBlanch** [comment]
Going to keep exemplars off by default for now based on performance analysis.
